persistence/src/main/scala/akka/persistence/LocalActorRef.scala

```diff
@@ -0,0 +1,134 @@
+import com.evolutiongaming.catshelper.{SerialRef, ToTry}
+import scala.concurrent.duration.FiniteDuration
+import java.time.Instant
+import java.util.concurrent.TimeoutException
+import java.time.temporal.ChronoUnit
+
+/** Representation of actor capable of constructing result from multiple messages passed into the actor. Inspired by [[PromiseActorRef]] but
+  * result [[R]] is an aggregate from incomming messages rather that first message. Can be used only locally, does _not_ tolerate.
+  * [[ActorRef.provider]] and [[ActorRef.path]] functions.
+  * @tparam F
+  *   The effect type.
+  * @tparam R
+  *   The result type of the aggregate.
+  */
+private[persistence] trait LocalActorRef[F[_], R] {
+
+  def ref: ActorRef
+
+  /** Semantically blocking while aggregating result
+    */
+  def res: F[R]
+
+  /** Immidiately get currect state:
+    *
+    * \- [[None]] if aggregating not finished yet
+    *
+    * \- [[Some(Left(Throwable))]] if aggregation failed or timeout happened
+    *
+    * \- [[Some(Right(r))]] if aggregation completed successfully
+    */
+  def get: F[Option[Either[Throwable, R]]]
+}
+
+private[persistence] object LocalActorRef {
+
+  type M = Any
+
+  /** Create new [[LocalActorRef]]
+    *
+    * @param initial
+    *   The initial state of type [[S]].
+    * @param timeout
+    *   [[TimeoutException]] will be thrown if no incomming messages received within the timeout.
+    * @param receive
+    *   The aggregate function defining how to apply incomming message on state or produce final result: [[Left]] for continue aggregating
+    *   while [[Right]] for the result.
+    * @tparam F
+    *   The effect type.
+    * @tparam S
+    *   The aggregating state type.
+    * @tparam R
+    *   The final result type.
+    * @return
+    */
+  def apply[F[_]: Temporal: ToTry, S, R](initial: S, timeout: FiniteDuration)(
+    receive: PartialFunction[(S, M), F[Either[S, R]]]
+  ): F[LocalActorRef[F, R]] = {
+
+    val F = Temporal[F]
+
+    case class State(state: S, updated: Instant)
+
+    def timeoutException = new TimeoutException(s"no messages received during period of $timeout")
+
+    for {
+      now   <- F.realTimeInstant
+      state <- SerialRef.of[F, State](State(initial, now))
+      defer <- F.deferred[Either[Throwable, R]]
+      fiber <- F.start {
+        val f = for {
+          _ <- F.sleep(timeout)
+          s <- state.get
+          n <- F.realTimeInstant
+          c  = s.updated.plus(timeout.toNanos, ChronoUnit.NANOS).isBefore(n)
+          _ <- if (c) defer.complete(timeoutException.asLeft) else F.unit
+        } yield c
+
+        ().tailRecM { _ =>
+          f.ifF(().asRight, ().asLeft)
+        }
+      }
+    } yield new LocalActorRef[F, R] {
+
+      private def done(e: Either[Throwable, R]) =
+        for {
+          _ <- defer.complete(e)
+          _ <- fiber.cancel
+        } yield {}
+
+      override def ref: ActorRef = new MinimalActorRef {
```

Comment:
why can't your actor just be `(f: Any => F[Unit]) => ActorRef` ?

Reply:
`ActorRef` trait is pretty huge with lots of methods to be implemented, I prefer to delegate it to `MinimalActorRef`

Reply:
I was more talking about that you logic of state/io composition all mixed inside of `MinimalActorRef` impl, where you basically just need to call IO.
This is not a strong opinion

Reply:
Ah, got your point. Yes, I can easily extract the logic into separate function, but I do not see much benefits of it.

On ref creation I defined few variables:
```
val state: SerialRef[F,State] = ???
val defer: Deferred[F,Either[Throwable,R]] = ???
val fiber: Fiber[F,Throwable,Unit] = ???
```

and all three of them used in _the function_ that now is defined in `override def !(m: M)(implicit sender: ActorRef): Unit`. I tend to keep this coupled and incapsulated inside `LocalActorRef` implementation due to code complexity and no much benefits of exposing it outside.

Taking into account my previous statement I can propose somethings like:
```
      private def update(m: M): F[Unit] =
        state
          .update { s =>
            if (receive.isDefinedAt(s.state -> m)) {

              for {
                t <- Temporal[F].realTimeInstant
                r <- receive(s.state -> m)
                s <- r match {
                  case Left(s)  => State(s, t).pure[F]
                  case Right(r) => done(r.asRight).as(s)
                }
              } yield s

            } else {
              s.pure[F]
            }
          }
          .handleErrorWith { e =>
            done(e.asLeft).void
          }

        override def !(m: M)(implicit sender: ActorRef): Unit =
          update(m).toTry.get
```

but IMO its not much difference from how its done now